third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1,35 +1,287 @@
+      getBlocksPerSM(threads_per_sm, threads_per_block, dev_prop->warpSize);
+  iop.gdimy = blocks_per_sm * device_multiprocessor_count;
+  const int64_t outer_iter_min = 8;
+  const int64_t gdimy_max = scheduler_utils::roundUpToN(
+      ceilDiv(outer_dim_numel, outer_iter_min), device_multiprocessor_count);
+  while (iop.gdimy > gdimy_max && blocks_per_sm > 1) {
+    blocks_per_sm -= 1;
+    iop.gdimy = blocks_per_sm * device_multiprocessor_count;
+  }
+
+  // Step-3, set OuterParams Iteration dim: outer_unroll_factor, bdimx, gdimy
+  // (already done)
+  iop.outer_unroll_factor = inner_dim_numel >= 4096 ? 4 : 2;
+  iop.bdimx = scheduler_utils::roundUpPow2Or8(
+      ceilDiv(inner_dim_numel / iop.outer_unroll_factor, iop.gdimy));
+
+  // Step-4, set OuterParams Reduction dim: bdimy.
+  iop.bdimy = threads_per_block / iop.bdimx;
+
+  // Step-5, special case, when inner_dim_numel <= 1024, bdimx is usually small
+  // after divide by inner_vect and inner_batch. In this case, bdimy is used to
+  // parallelize outer_dim instead of inner_dim. This pattern is named multi
+  // reductions per block (mrpb).
+  if (inner_dim_numel <= 1024) {
+    rparams->multiple_reds_per_blk = true;
+    rparams->tidx_for_outer_reduction = true;
+    constexpr int64_t threads_per_block_mrpb = 512;
+
+    // Step-1, InnerParams, Reduction dim: inner_vect(reuse), inner_batch, bdimx
+    iop.inner_batch = 1;
+    iop.bdimx = inner_dim_numel / iop.inner_vect;
+
+    // Step-2, InnerParams, Iteration dim: gdimy, bdimy (in next step)
+    reg_per_thread =
+        getEstimatedRegisterUsage(iop.inner_vect * iop.inner_batch);
+    threads_per_sm = getThreadsPerSMGivenRegPerThread(reg_per_thread);
+    blocks_per_sm = getBlocksPerSM(
+        threads_per_sm, threads_per_block_mrpb, dev_prop->warpSize);
+    iop.gdimy = blocks_per_sm * device_multiprocessor_count;
+
+    // Step-3, OuterParams, Iteration dim: outer_unroll_factor, bdimy, gdimy (in
+    // previous step). unroll_factor is set to 2 as we only process vectorized
+    // cases.
+    iop.outer_unroll_factor = 2;
+    iop.bdimy = std::min(
+        ceilDiv(inner_dim_numel / iop.outer_unroll_factor, iop.gdimy),
+        scheduler_utils::safeDiv(threads_per_block_mrpb, iop.bdimx));
+    iop.bdimy = iop.bdimy;
+
+    // Step-4, OuterParams, Reduction dim: bdimx (already done)
+
+    if (iop.bdimx % dev_prop->warpSize == 0) {
+      rparams->pad_inner_reduction_to_warp = true;
+      rparams->pad_outer_reduction_to_warp = true;
+    }
+    rparams->block_dim_iter_dom = ParallelType::TIDy;
+  } else {
+    rparams->block_dim_inner_reduction_extra = ParallelType::TIDy;
+  }
+
+  // The partial outer reduction result is stored in tmp gmem, set the
+  // vectorization factor for write and read
+  constexpr int64_t max_gmem_vect_access_bytes = 16;
+  const int64_t max_tmp_gmem_vect_factor =
+      max_gmem_vect_access_bytes / tmp_gmem_dtype_size;
+  iop.tmp_gmem_write_vect = std::min(max_tmp_gmem_vect_factor, iop.inner_vect);
```

Comment:
Is it possible to have an  inner vectorization factor that is different from the write vectorization factor of the tmp buffer? I wonder if they should always be the same.

Reply:
There is a global sync after write and we changed parallel pattern for read, so the vect factor can be different.

Reply:
Not quite sure. I understand there's a global sync after writes to the temp buffer, but what I asked was the inner vectorization factor and the vectorization factor of the writes to the temp buffer. Both vectorizations should occur before the global sync, right? If so, why does the global sync matter?

Reply:
yes. they can be different and are different in cases where input data type is `half` and vectorized by 8, the wirte to temp buffer is vectorized by 4. Each thread will do the write twice each with 4 floats.

Reply:
Just to confirm, does the global sync have anything to do with? 

I have a related question here https://github.com/csarofeen/pytorch/pull/2400/files#r1152631408

Reply:
global sync has nothing to do with ` inner vectorization factor ` and `write vectorization factor of the tmp buffer`.